tests/utils/exec.ts

```diff
@@ -48,31 +62,40 @@ const generateDockerArgs = (mountPath: string) => [
 ];
 
 /**
- * Wrapper for execFile to make it asynchronous and non-blocking
+ * Execute general (non-Polykey) commands
  */
 async function exec(
   command: string,
   args: Array<string> = [],
+  opts: ExecOpts = { env: {} },
 ): Promise<{
   stdout: string;
   stderr: string;
 }> {
+  const env = {
+    ...process.env,
+    ...opts.env,
+  };
   return new Promise((resolve, reject) => {
-    child_process.execFile(
-      command,
-      args,
-      { windowsHide: true },
-      (error, stdout, stderr) => {
-        if (error) {
-          reject(error);
-        } else {
-          return resolve({
-            stdout,
-            stderr,
-          });
-        }
-      },
-    );
+    let stdout = '',
+      stderr = '';
+    const subprocess = childProcess.spawn(command, args, {
+      env,
+      windowsHide: true,
+      shell: opts.shell ? opts.shell : false,
+    });
+    subprocess.stdout.on('data', (data) => {
+      stdout += data.toString();
+    });
+    subprocess.stderr.on('data', (data) => {
+      stderr += data.toString();
+    });
```

Comment:
Where should the exit code go?

Reply:
I suppose we could return the code, but since this is just for running shell commands the code itself probably wouldn't be too informative. Wouldn't it just be 0 or 1 for the most part?

Reply:
If this is intended to return a `Promise`, you need to capture the error and throw it as a promise if the process failed to spawn.

A successful exit code is always `0`. See: https://nodejs.org/api/child_process.html#event-error.

And also:

> If this method is invoked as its [util.promisify()](https://nodejs.org/api/util.html#utilpromisifyoriginal)ed version, it returns a Promise for an Object with stdout and stderr properties. The returned ChildProcess instance is attached to the Promise as a child property. In case of an error (including any error resulting in an exit code other than 0), a rejected promise is returned, with the same error object given in the callback, but with two additional properties stdout and stderr.

Reply:
So no need to return the `exitCode`. It should be part of the child process exception.

Can you find out what the callback returns in the case of a failing call. I think it's something like `ChildProcessError`. Prototype this.

Reply:
I just checked the old `pkStdio`, it does in fact return the `exitCode`, but this is only because it's only executing a function. As long as it is doing that, returning the `exitCode` should be done.

However if you are executing a child process, any command that doesn't exist with `0` is considered an error right?

Reply:
The new `pkStdio` returns the exit code. This is for `exec` which is only intended to run shell commands, not Polykey commands. So there is no `exitCode` property, only `code`.

Reply:
Yea, I know but we want the APIs to be similar. So the question is, what does spawn's error event look like and does it cover non-0 exit codes?

Reply:
It looks like the `error` event is for higher-level things to do with starting/stopping/communicating with the child process https://nodejs.org/api/child_process.html#event-error - it returns a plain old error object. The `exit` event returns an exit code and signal https://nodejs.org/api/child_process.html#event-exit.

Reply:
If that's the case, then the error event should result in you `reject(e)`, and then for non-0 exit code, you just need to resolve with the `exitCode` as a returned property.

I think there's actually a similar example of this already being done elsewhere.

Reply:
Make sure to cleanly separate things like:

* `testUtils.exec` - just executes things
* `testUtils.pkExec` - executes PK

If we need a generic spawn then:

* `testUtils.spawn` - generic spawning
* `testUtils.pkSpawn` - specifically pk spawning

All functions end up using `childProcess.spawn`.

Reply:
We now have a generic `exec` and generic `spawn`. The NAT utils no longer need to use `child_process` directly and now use either `testUtils.exec` or `testUtils.spawn` when calling generic shell commands.

Reply:
Great. @tegefaulkes we should then avoid using `childProcess` directly (only for its types with `import type { ... } from 'child_process';`.

Any executions goes to `testUtils.spawn` or `testUtils.exec`.